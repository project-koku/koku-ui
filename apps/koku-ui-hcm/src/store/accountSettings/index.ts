import * as accountSettingsActions from './accountSettingsActions';
import { accountSettingsStateKey } from './accountSettingsCommon';
import type { AccountSettingsAction, AccountSettingsState } from './accountSettingsReducer';
import { accountSettingsReducer } from './accountSettingsReducer';
import * as accountSettingsSelectors from './accountSettingsSelectors';

export type { AccountSettingsAction, AccountSettingsState };
export { accountSettingsActions, accountSettingsReducer, accountSettingsSelectors, accountSettingsStateKey };
