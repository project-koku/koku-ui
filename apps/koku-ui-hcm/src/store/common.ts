import type { MapDispatchToProps, MapStateToProps } from 'react-redux';
import type { Action } from 'redux';
import type { ThunkAction as ReduxThunkAction } from 'redux-thunk';

import type { RootState } from './rootReducer';

export const expirationMS = 30 * 60 * 1000; // 30 minutes

export const enum FetchStatus {
  'none',
  'inProgress',
  'complete',
}

export function createMapStateToProps<OwnProps, StateProps>(
  mapStateToProps: MapStateToProps<StateProps, OwnProps, RootState>
) {
  return mapStateToProps;
}

export function createMapDispatchToProps<OwnProps, DispatchProps>(
  mapDispatchToProps: MapDispatchToProps<DispatchProps, OwnProps>
) {
  return mapDispatchToProps;
}

export type ThunkAction<A extends Action = any, R = any> = ReduxThunkAction<R, RootState, void, A>;
