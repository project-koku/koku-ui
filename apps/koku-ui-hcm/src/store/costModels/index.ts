import * as costModelsActions from './actions';
import type { CostModelsAction, CostModelsState } from './reducer';
import { reducer as costModelsReducer, stateKey as costModelsStateKey } from './reducer';
import * as costModelsSelectors from './selectors';

export type { CostModelsAction, CostModelsState };
export { costModelsActions, costModelsReducer, costModelsSelectors, costModelsStateKey };
