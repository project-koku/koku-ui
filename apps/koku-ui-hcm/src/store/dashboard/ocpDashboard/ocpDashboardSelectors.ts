import type { RootState } from 'store/rootReducer';
import { getCurrency } from 'utils/sessionStorage';

import {
  getQueryForWidget,
  getQueryForWidgetTabs,
  ocpDashboardDefaultFilters,
  ocpDashboardStateKey,
  OcpDashboardTab,
  ocpDashboardTabFilters,
} from './ocpDashboardCommon';

export const selectOcpDashboardState = (state: RootState) => state[ocpDashboardStateKey];

export const selectWidgets = (state: RootState) => selectOcpDashboardState(state).widgets;

export const selectWidget = (state: RootState, id: number) => selectWidgets(state)[id];

export const selectCurrentWidgets = (state: RootState) => selectOcpDashboardState(state).currentWidgets;

export const selectWidgetQueries = (state: RootState, id: number) => {
  const widget = selectWidget(state, id);

  const defaultFilter = {
    ...ocpDashboardDefaultFilters,
    ...(widget.filter ? widget.filter : {}),
  };
  const tabsFilter = {
    ...ocpDashboardTabFilters,
    ...(widget.tabsFilter ? widget.tabsFilter : {}),
  };
  const props = {
    currency: getCurrency(),
  };
  const tabsProps = {
    ...props,
    ...(widget.trend &&
      widget.trend.costDistribution &&
      widget.currentTab === OcpDashboardTab.projects && {
        order_by: {
          distributed_cost: 'desc',
        },
      }),
  };

  return {
    previous: getQueryForWidget(
      {
        ...defaultFilter,
        time_scope_value: -2,
      },
      props
    ),
    current: getQueryForWidget(defaultFilter, props),
    forecast: getQueryForWidget({}, { limit: 31, ...props }),
    optimizations: getQueryForWidget({}),
    tabs: getQueryForWidgetTabs(
      widget,
      {
        ...(tabsFilter as any),
        resolution: 'monthly',
      },
      tabsProps
    ),
  };
};
