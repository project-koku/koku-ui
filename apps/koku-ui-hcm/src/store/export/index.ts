import * as exportActions from './exportActions';
import { exportStateKey } from './exportCommon';
import type { ExportAction, ExportState } from './exportReducer';
import { exportReducer } from './exportReducer';
import * as exportSelectors from './exportSelectors';

export { exportActions, exportReducer, exportSelectors, exportStateKey };
export type { ExportAction, ExportState };
