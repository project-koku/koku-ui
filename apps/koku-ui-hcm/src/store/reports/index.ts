import * as reportActions from './reportActions';
import { reportStateKey } from './reportCommon';
import type { CachedReport, ReportAction, ReportState } from './reportReducer';
import { reportReducer } from './reportReducer';
import * as reportSelectors from './reportSelectors';

export type { ReportAction, CachedReport, ReportState };
export { reportActions, reportReducer, reportSelectors, reportStateKey };
