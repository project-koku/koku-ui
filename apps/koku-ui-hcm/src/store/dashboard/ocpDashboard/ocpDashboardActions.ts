import type { ThunkAction } from 'store/common';
import { forecastActions } from 'store/forecasts';
import { reportActions } from 'store/reports';
import { createAction } from 'typesafe-actions';

import type { OcpDashboardTab } from './ocpDashboardCommon';
import { selectWidget, selectWidgetQueries } from './ocpDashboardSelectors';

export const fetchWidgetForecasts = (id: number): ThunkAction => {
  return (dispatch, getState) => {
    const state = getState();
    const widget = selectWidget(state, id);

    if (widget.forecastPathsType && widget.forecastType) {
      const { forecast } = selectWidgetQueries(state, id);
      dispatch(forecastActions.fetchForecast(widget.forecastPathsType, widget.forecastType, forecast));
    }
  };
};

export const fetchWidgetReports = (id: number): ThunkAction => {
  return (dispatch, getState) => {
    const state = getState();
    const widget = selectWidget(state, id);
    const { previous, current, tabs } = selectWidgetQueries(state, id);

    if (widget.reportPathsType && widget.reportType) {
      dispatch(reportActions.fetchReport(widget.reportPathsType, widget.reportType, current));
      dispatch(reportActions.fetchReport(widget.reportPathsType, widget.reportType, previous));
      if (widget.availableTabs) {
        dispatch(reportActions.fetchReport(widget.reportPathsType, widget.reportType, tabs));
      }
    }
  };
};

export const setWidgetTab = createAction('ocpDashboard/widget/tab')<{
  id: number;
  tab: OcpDashboardTab;
}>();

export const changeWidgetTab = (id: number, tab: OcpDashboardTab): ThunkAction => {
  return dispatch => {
    dispatch(setWidgetTab({ id, tab }));
    dispatch(fetchWidgetReports(id));
  };
};
