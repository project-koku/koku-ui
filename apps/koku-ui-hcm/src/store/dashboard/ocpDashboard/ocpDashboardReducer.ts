import type { DashboardWidget } from 'store/dashboard/common/dashboardCommon';
import type { ActionType } from 'typesafe-actions';
import { getType } from 'typesafe-actions';

import { setWidgetTab } from './ocpDashboardActions';
import { costSummaryWidget, cpuWidget, memoryWidget, optimizationsWidget, volumeWidget } from './ocpDashboardWidgets';

export type OcpDashboardAction = ActionType<typeof setWidgetTab>;

export type OcpDashboardState = Readonly<{
  widgets: Record<number, DashboardWidget>;
  currentWidgets: number[];
}>;

export const defaultState: OcpDashboardState = {
  currentWidgets: [costSummaryWidget.id, cpuWidget.id, memoryWidget.id, volumeWidget.id, optimizationsWidget.id],
  widgets: {
    [costSummaryWidget.id]: costSummaryWidget,
    [cpuWidget.id]: cpuWidget,
    [memoryWidget.id]: memoryWidget,
    [volumeWidget.id]: volumeWidget,
    [optimizationsWidget.id]: optimizationsWidget,
  },
};

export function ocpDashboardReducer(state = defaultState, action: OcpDashboardAction): OcpDashboardState {
  switch (action.type) {
    case getType(setWidgetTab):
      return {
        ...state,
        widgets: {
          ...state.widgets,
          [action.payload.id]: {
            ...state.widgets[action.payload.id],
            currentTab: action.payload.tab,
          },
        },
      };
    default:
      return state;
  }
}
