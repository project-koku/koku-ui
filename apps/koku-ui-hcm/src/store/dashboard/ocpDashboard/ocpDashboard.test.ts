jest.mock('store/reports/reportActions');

import { ReportType } from 'api/reports/report';
import { ComputedReportItemType, DatumType } from 'routes/components/charts/common/chartDatum';
import { createMockStoreCreator } from 'store/mockStore';
import { reportActions } from 'store/reports';

import * as actions from './ocpDashboardActions';
import { getGroupByForTab, getQueryForWidgetTabs, ocpDashboardStateKey, OcpDashboardTab } from './ocpDashboardCommon';
import { ocpDashboardReducer } from './ocpDashboardReducer';
import * as selectors from './ocpDashboardSelectors';
import { costSummaryWidget, cpuWidget, memoryWidget, optimizationsWidget, volumeWidget } from './ocpDashboardWidgets';

const createOcpDashboardStore = createMockStoreCreator({
  [ocpDashboardStateKey]: ocpDashboardReducer,
});

const fetchReportMock = reportActions.fetchReport as jest.Mock;

beforeEach(() => {
  fetchReportMock.mockReturnValue({ type: '@@test' });
});

test('default state', () => {
  const store = createOcpDashboardStore();
  const state = store.getState();
  expect(selectors.selectCurrentWidgets(state)).toEqual([
    costSummaryWidget.id,
    cpuWidget.id,
    memoryWidget.id,
    volumeWidget.id,
    optimizationsWidget.id,
  ]);
  expect(selectors.selectWidget(state, costSummaryWidget.id)).toEqual(costSummaryWidget);
});

test('fetch widget reports', () => {
  const store = createOcpDashboardStore();
  store.dispatch(actions.fetchWidgetReports(costSummaryWidget.id));
  expect(fetchReportMock.mock.calls).toMatchSnapshot();
});

test('changeWidgetTab', () => {
  const store = createOcpDashboardStore();
  store.dispatch(actions.changeWidgetTab(costSummaryWidget.id, OcpDashboardTab.projects));
  const widget = selectors.selectWidget(store.getState(), costSummaryWidget.id);
  expect(widget.currentTab).toBe(OcpDashboardTab.projects);
  expect(fetchReportMock).toHaveBeenCalledTimes(3);
});

describe('getGroupByForTab', () => {
  test('clusters tab', () => {
    expect(getGroupByForTab(OcpDashboardTab.clusters)).toMatchSnapshot();
  });

  test('nodes tab', () => {
    expect(getGroupByForTab(OcpDashboardTab.nodes)).toMatchSnapshot();
  });

  test('projects tab', () => {
    expect(getGroupByForTab(OcpDashboardTab.projects)).toMatchSnapshot();
  });

  test('unknown tab', () => {
    expect(getGroupByForTab('unknown' as any)).toMatchSnapshot();
  });
});

test('getQueryForWidget', () => {
  const widget = {
    id: 1,
    titleKey: '',
    reportType: ReportType.cost,
    availableTabs: [OcpDashboardTab.projects],
    currentTab: OcpDashboardTab.projects,
    details: { formatOptions: {} },
    trend: {
      computedReportItem: ComputedReportItemType.cost,
      datumType: DatumType.rolling,
      formatOptions: {},
      titleKey: '',
    },
    topItems: {
      formatOptions: {},
    },
  };

  [
    [
      undefined,
      'filter[resolution]=daily&filter[time_scope_units]=month&filter[time_scope_value]=-1&group_by[project]=*',
    ],
    [{}, 'group_by[project]=*'],
    [{ limit: 3 }, 'filter[limit]=3&group_by[project]=*'],
  ].forEach(value => {
    expect(getQueryForWidgetTabs(widget, value[0])).toEqual(value[1]);
  });
});
