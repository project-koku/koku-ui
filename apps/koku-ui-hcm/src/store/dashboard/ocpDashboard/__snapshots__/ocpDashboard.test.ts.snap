// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`fetch widget reports 1`] = `
[
  [
    "ocp",
    "cost",
    "currency=USD&filter[resolution]=daily&filter[time_scope_units]=month&filter[time_scope_value]=-1",
  ],
  [
    "ocp",
    "cost",
    "currency=USD&filter[resolution]=daily&filter[time_scope_units]=month&filter[time_scope_value]=-2",
  ],
  [
    "ocp",
    "cost",
    "currency=USD&filter[limit]=3&filter[resolution]=monthly&filter[time_scope_units]=month&filter[time_scope_value]=-1&group_by[project]=*&order_by[distributed_cost]=desc",
  ],
]
`;

exports[`getGroupByForTab clusters tab 1`] = `
{
  "cluster": "*",
}
`;

exports[`getGroupByForTab nodes tab 1`] = `
{
  "node": "*",
}
`;

exports[`getGroupByForTab projects tab 1`] = `
{
  "project": "*",
}
`;

exports[`getGroupByForTab unknown tab 1`] = `{}`;
